concordium-std/src/types.rs

```diff
@@ -384,7 +384,12 @@ impl<'a, V> crate::ops::Deref for StateRef<'a, V> {
 /// that the value is properly stored in the contract state maintained by the
 /// node.
 pub struct StateRefMut<'a, V: Serial, S: HasStateApi> {
-    pub(crate) entry:            UnsafeCell<S::EntryType>,
+    /// This is set as an `UnsafeCell`, to be able to get a mutable reference to
+    /// the entry without `StateRefMut` being mutable.
+    /// The `Option` allows for having an internal method destroying the
+    /// `StateRefMut` into its raw parts without `Drop` causing a write to the
+    /// contract state.
```

Comment:
I don't really understand this, and I missed where this was actually needed. Can you point?

Also, why is std::mem::forget not a good alternative?

Reply:
It is needed in the `StateRefMut::into_raw_parts` method that I have introduced.
The point is to be able to take out the value and entry of the `StateRefMut` while preventing it from writing mutations to the key-value store.

`std::mem::forget` takes ownership of the value, so it won't be dropped, but we lose access to it as well in that case.

Reply:
I see, you have problems moving out of it because we have a manual Drop implementation, but as far as I see this is actually just used once in delete_node.

Would it be very inconvenient to just directly call prims::state_delete_entry?

It's annoying to make the common cases worse for StateRefMut to work around a type system limitation.

Reply:
I'll try to avoid introducing the option, maybe replace `into_raw_parts` with something for deleting the entry directly and dropping the `StateRefMut` instead.

Reply:
Calling `prims::state_delete_entry` would work, but it means that this data structure won't work with the old infrastructure. If we decide that is okay, then we might as well remove the type parameter `S` and hard-code `StateApi` for `StateBTree*`.

Another, more involved solution is to introduce a separate `StateRefMut`-like type just for this.

Reply:
We discussed this offline, and the conclusion was that since we have deprecated the old unit testing infrastructure, we don't have to support it for this data structure.